README.md
optimize.js
===========

[This](http://en.wikipedia.org/wiki/Mathematical_optimization) kind of
optimization. Not [that](https://developers.google.com/closure/compiler/).

This was ported from the `scipy.optimize` and we seem to be about an order
of magnitude faster than the Python version but that hasn't been too well
tested.

Usage
-----

To optimize some dumb function, just run

```javascript
my_function = function (x) {
    return optimize.vector.dot(x, x);
};
xopt = optimize.fmin(my_function, [5.0, -3.4, 1.7, 16.3, 0.17]);
```

And this should say something like:

```
Converged in 349 iterations.
Function value = 3.954810202072493e-7
```

And `xopt` should end up being something like

```
[-0.0003993727670733724, -0.00027257793115893254, -0.0003811443958917447, 0.00012560283306980614, 0.00002523018782488204]
```

You can also include some options as follows:

```javascript
xopt = optimize.fmin(my_function, [5.0, -3.4, 1.7, 16.3, 0.17], {ftol: 1e-7, maxiter: 1000});
```

To Do
-----

* Write tests.
* Implement some better algorithms.
* Make some demos.


CHANGELOG.md

# 0.1.0 (2014-11-10)

- Initial version
Based off of https://github.com/dfm/optimize.js

I have made improvements to comply with the jshintrc
and made it a node module


test/vector_test.js
'use strict';

var vector = require('../lib/vector');
var assert = require('should');

var chai = require('chai');
var expect = chai.expect;

describe('vector', function () {
    it('arg sort return self', function () {
        vector.argsort((1)).should.equal(1);
    });

    describe('coverage', function(){
        it('dot', function(){
            var fn = function(){
                vector.dot([1], [1,1]);
            };

            expect(fn).to.throw('Size mismatch in vector.dot.');
        });

        it('add', function(){
            var fn = function(){
                vector.add([1], [1,1]);
            };

            expect(fn).to.throw('Size mismatch in vector.add.');
        });

        it('subtract', function(){
            var fn = function(){
                vector.subtract([1], [1,1]);
            };

            expect(fn).to.throw('Size mismatch in vector.subtract.');
        });

        it('copy', function(){
            vector.copy(1).should.equal(1);
        });
    });

    it('shape', function(){
        vector.shape([[1, 2], [1, 2]]).should.eql([2, 2]);
    })

});



test/optimize_test.js
'use strict';

var optimize = require('../');
var assert = require('should');

describe('neldermead', function () {
    it('single independent var', function () {
        var ans = optimize.fmin(function(x){
            return 3 * Math.pow(x - 2, 2);
        }, 0);

        ans.x.should.eql([ 1.9997500000000024 ]);
    });

    it('multiple independent var', function () {
        var ans = optimize.fmin(function(x){
            return Math.pow(x[0] * x[1] - 10, 2);
        }, [0.5, 0.5], {
            ftol: 1e-11,
            ep: 1.5e-9,
            maxiter: 500
        });

        ans.x.should.eql([ 3.1607583999633713, 3.163797950744621 ] );
    });
});

describe('newton', function () {
    it('single independent var', function () {
        var ans = optimize.newton(function(x){
            return [3 * Math.pow(x - 2, 2),3 * Math.pow(x - 2, 2)];
        }, 0, {
            ftol: 1e-11,
            ep: 1.5e-9,
            maxiter: 500
        });

        if(Math.abs(1.9995117262371358 - ans) > 0.1){
            throw 'not close enough';
        }
    });
});


gulpfile.js
'use strict';

var gulp   = require('gulp');
var plugins = require('gulp-load-plugins')();

var paths = {
    lint: ['./gulpfile.js', './lib/**/*.js'],
    watch: ['./gulpfile.js', './lib/**', './test/**/*.js', '!test/{temp,temp/**}'],
    tests: ['./test/**/*.js', '!test/{temp,temp/**}'],
    source: ['./lib/*.js']
};

var onError = function(err) {
    plugins.util.beep();

    if (process.env.CI) {
        throw new Error(err);
    }
};

gulp.task('lint', function () {
    return gulp.src(paths.lint)
        .pipe(plugins.jshint('.jshintrc'))
        .pipe(plugins.plumber({
            errorHandler: onError
        }))
        .pipe(plugins.jscs())
        .pipe(plugins.jshint.reporter('jshint-stylish'));
});

gulp.task('istanbul', function (cb) {
    gulp.src(paths.source)
        .pipe(plugins.istanbul()) // Covering files
        .on('finish', function () {
            gulp.src(paths.tests)
                .pipe(plugins.plumber({
                    errorHandler: onError
                }))
                .pipe(plugins.mocha())
                .pipe(plugins.istanbul.writeReports()) // Creating the reports after tests runned
                .on('finish', function() {
                    process.chdir(__dirname);
                    cb();
                });
        });
});

gulp.task('bump', ['test'], function () {
    var bumpType = plugins.util.env.type || 'patch'; // major.minor.patch

    return gulp.src(['./package.json'])
        .pipe(plugins.bump({ type: bumpType }))
        .pipe(gulp.dest('./'));
});

gulp.task('watch', ['test'], function () {
    gulp.watch(paths.watch, ['test']);
});

gulp.task('test', ['lint', 'istanbul']);

gulp.task('release', ['bump']);

gulp.task('default', ['test']);


lib/vector.js
var vector = module.exports = {};

function merge(l, r, data) {
    // Merging for use with a merge argsort.
    var result = [];
    while (l.length && r.length) {
        if (data[l[0]] <= data[r[0]]){
            result.push(l.shift());
        }else{
            result.push(r.shift());
        }
    }
    while (l.length){
        result.push(l.shift());
    }
    while (r.length){
        result.push(r.shift());
    }
    return result;
}

vector.shape = function(arr, ret){
    if(!ret){ ret = []; }

    if(Array.isArray(arr)){
        ret.push(arr.length);
        return vector.shape(arr[0], ret);
    }else{
        return ret;
    }
};

vector.copy = function (x) {
    var y, i;
    if (typeof(x.length) === "undefined"){
        return x;
    }

    y = [];
    for (i = 0; i < x.length; i++){
        y[i] = x[i];
    }

    return y;
};

vector.atLeast1d = function (x) {
    // Make sure that an object acts as an array (even if it's a scalar).
    if (typeof(x.length) === "undefined") {
        var tmp = [];
        tmp[0] = x;
        return tmp;
    }
    return x;
};

vector.range = function (a, b, c) {
    var xmin, xmax, dx, x, rng = [];
    if (typeof(b) === "undefined") {
        xmin = 0;
        xmax = a;
        dx = 1;
    } else if (typeof(c) === "undefined") {
        xmin = a;
        xmax = b;
        dx = 1;
    } else {
        xmin = a;
        xmax = b;
        dx = c;
    }

    var i;
    for (x = xmin, i = 0; x < xmax; x += dx, i++){
        rng[i] = x;
    }

    return rng;
};

vector.dot = function (a, b) {
    var i, result = 0.0;
    if (a.length !== b.length){
        throw "Size mismatch in vector.dot.";
    }

    for (i = 0; i < a.length; i++) {
        result += a[i] * b[i];
    }
    return result;
};

vector.fmult = function (f, v) {
    var i, result = [];
    for (i = 0; i < v.length; i++) {
        result[i] = f * v[i];
    }
    return result;
};

vector.add = function (a, b) {
    var i, result = [];
    if (a.length !== b.length) {
        throw "Size mismatch in vector.add.";
    }
    for (i = 0; i < a.length; i++) {
        result[i] = a[i] + b[i];
    }
    return result;
};

vector.subtract = function (a, b) {
    var i, result = [];
    if (a.length !== b.length) {
        throw "Size mismatch in vector.subtract.";
    }
    for (i = 0; i < a.length; i++) {
        result[i] = a[i] - b[i];
    }
    return result;
};

vector.reduce = function (x) {
    var i, result = x[0];
    for (i = 1; i < x.length; i++){
        result = vector.add(result, x[i]);
    }
    return result;
};

vector.take = function (x, ind) {
    // Re-order a vector.
    var i, result = [];
    for (i = 0; i < ind.length; i++) {
        result[i] = x[ind[i]];
    }
    return result;
};

function rec_argsort(inds, data) {
    // The recursive helper function for the argsort.
    var m, l, r;
    if (typeof(inds) === "undefined" || inds.length === 1) {
        return inds;
    }

    m = parseInt(inds.length / 2);
    l = inds.slice(0, m);
    r = inds.slice(m, inds.length);
    return merge(rec_argsort(l, data),
        rec_argsort(r, data), data);
}

vector.argsort = function (x) {
    // Argsort of an array using merge sort.
    if (typeof(x.length) === "undefined" || x.length === 1) {
        return x;
    }

    return rec_argsort(vector.range(x.length), x);
};


lib/optimize.js
var vector = require('./vector');
var sylvester = require('sylvester');

var $V = sylvester.Vector.create;
var $M = sylvester.Matrix.create;

var optimize = module.exports = {};

function q(x) {
    // The existential operator;
    return typeof(x) !== "undefined" && x !== null;
}

function approx_jacobian(x, f, ep) {
    // Approximate NxM dimensional gradient of the vector function
    // f using forward finite difference.
    var f0 = f(x);
    var grad = [];
    var eps = [];

    x = vector.atLeast1d(x);
    var x0 = vector.copy(x);

    if (typeof(ep.length) === "undefined") {
        for (var i = 0; i < x.length; i++){
            eps.push(ep);
        }
    } else if (ep.length === x.length) {
        eps = ep;
    } else {
        throw "Size mismatch in _approx_fprime.";
    }

    for (var k = 0; k < x.length; k++) {
        x[k] = x0[k] + eps[k];
        grad[k] = $V(f(x)).subtract(f0).x(1.0 / eps[k]).elements;
        x[k] = x0[k];
    }

    return $M(grad).transpose();
}

function max_abs_diff(x0, x) {
    var max = 0.0;

    for (var i = 0; i < x.length; i++){
        max = Math.max(max, Math.abs(x0 - x[i]));
    }
    return max;
}

optimize.fmin = function (func, x0, opts) {
    // Optimize a function using Nelder-Mead.

    x0 = vector.atLeast1d(x0);
    var N = x0.length;

    // Defaults.
    if (!q(opts)){ opts = {}; }

    // Not used
    //var xtol = q(opts.xtol) ? opts.xtol : 1e-6;

    var ftol = q(opts.ftol) ? opts.ftol : 1e-6;
    var maxiter = q(opts.maxiter) ? opts.maxiter : 200 * N;

    // Magic numbers from `scipy`.
    var rho = 1;
    var chi = 2;
    var psi = 0.5;
    var sigma = 0.5;
    var nonzdelt = 0.05;
    var zdelt = 0.00025;

    var sim = [];
    sim[0] = x0;
    var fsim = [];
    fsim[0] = func(x0);

    for (var i = 0; i < N; i++) {
        var y = vector.copy(x0);
        if (y[i] !== 0.0){
            y[i] *= (1 + nonzdelt);
        }else {
            y[i] = zdelt;
        }

        sim[i + 1] = y;
        fsim[i + 1] = func(y);
    }

    var inds = vector.argsort(fsim);
    fsim = vector.take(fsim, inds);
    sim = vector.take(sim, inds);

    var iterations = 0;

    // Constraint on function calls is needed.
    while (iterations < maxiter) {
        var doshrink = false;

        iterations += 1;
        // A break based on xtol needs to be included too.
        if (max_abs_diff(fsim[0], fsim.slice(1, fsim.length)) <= ftol){
            break;
        }

        var xbar = vector.fmult(1.0 / N,
            vector.reduce(sim.slice(0, sim.length - 1)));
        var xr = vector.add(vector.fmult(1 + rho, xbar),
            vector.fmult(-rho, sim[sim.length - 1]));
        var fxr = func(xr);

        if (fxr < fsim[0]) {
            var xe = vector.add(vector.fmult(1 + rho * chi, xbar),
                vector.fmult(-rho * chi, sim[sim.length - 1]));
            var fxe = func(xe);
            if (fxe < fxr) {
                sim[sim.length - 1] = xe;
                fsim[fsim.length - 1] = fxe;
            } else {
                sim[sim.length - 1] = xr;
                fsim[fsim.length - 1] = fxr;
            }
        } else {
            if (fxr < fsim[fsim.length - 2]) {
                sim[sim.length - 1] = xr;
                fsim[fsim.length - 1] = fxr;
            } else {
                var xc, fxc;
                if (fxr < fsim[fsim.length - 1]) {
                    xc = vector.add(vector.fmult(1 + rho * psi, xbar),
                        vector.fmult(-rho * psi, sim[sim.length - 1]));
                    fxc = func(xc);
                    if (fxc < fsim[fsim.length - 1]) {
                        sim[sim.length - 1] = xc;
                        fsim[fsim.length - 1] = fxc;
                    } else {
                        doshrink = true;
                    }
                } else {
                    xc = vector.add(vector.fmult(1 - psi, xbar),
                        vector.fmult(psi, sim[sim.length - 1]));
                    fxc = func(xc);
                    if (fxc < fsim[fsim.length - 1]) {
                        sim[sim.length - 1] = xc;
                        fsim[fsim.length - 1] = fxc;
                    } else {
                        doshrink = true;
                    }
                }

                if (doshrink) {
                    for (var j = 1; j < N + 1; j++) {
                        sim[j] = vector.add(sim[0], vector.fmult(sigma,
                            vector.subtract(sim[j], sim[0])));
                        fsim[j] = func(sim[j]);
                    }
                }
            }
        }

        inds = vector.argsort(fsim);
        fsim = vector.take(fsim, inds);
        sim = vector.take(sim, inds);
    }

    var x = sim[0];
    var fval = fsim[0];

    var result = {};

    if (iterations >= maxiter){
        result.error = "Too many iterations. " + iterations;
    }else{
        result.result = "Converged in " + iterations + " iterations.";
    }

    result.x = x;
    result.fval = fval;

    return result;
};

optimize.newton = function (fn, x0, opts) {
    // fn should return the chi vector (data - model) / sigma.
    // Also, this function uses _dumb-ass_ inversion. We suck.
    x0 = vector.atLeast1d(x0);
    var chi0 = fn(x0);
    var chi20 = vector.dot(chi0, chi0);

    // Defaults.
    if (!q(opts)){ opts = {}; }

    var ftol = q(opts.ftol) ? opts.ftol : 1e-10;
    var ep = q(opts.ep) ? opts.ep : 1.49e-8;  // Magic number from scipy.
    var maxiter = q(opts.maxiter) ? opts.maxiter : 200 * x0.length;

    // Not used
    //var fprime = q(opts.fprime) ? opts.fprime : function (x) {
    //    return optimize._approx_fprime(x, fn, ep);
    //};

    var alpha = 1.0;

    for (var i = 0; i < maxiter; i++) {
        var J = approx_jacobian(x0, fn, ep);
        var JT = J.transpose();
        var JTJ = JT.x(J);

        // not used
        //var diagJTJ = sylvester.Matrix.Diagonal(JTJ.diagonal().elements);

        var JTfx = JT.x($V(chi0));

        var dx = JTJ.inv().x(JTfx);
        // dx = JTJ.add(diagJTJ.x(lambda)).inv().x(JTfx);

        var x_best = vector.copy(x0);
        var chi_best = vector.copy(chi0);
        var chi2_best = chi20;

        for (var n = 0; n <= 5; n++) {
            alpha = Math.pow(2, -n);

            var x_try = vector.subtract(x0, dx.x(alpha).elements);
            var chi_try = fn(x_try);
            var chi2_try = vector.dot(chi_try, chi_try);

            if (chi2_try < chi2_best) {
                x_best = x_try;
                chi_best = chi_try;
                chi2_best = chi2_try;
            }
        }

        var dchi2 = chi20 - chi2_best;

        x0 = x_best;
        chi0 = chi_best;
        chi20 = chi2_best;

        if (dchi2 < 0.0){
            throw "Failure";
        }

        if (i > 1 && dchi2 < ftol){
            break;
        }
    }

    return x0;
};


optimize-browser.js
(function e(t,n,r){function s(o,u){if(!n[o]){if(!t[o]){var a=typeof require=="function"&&require;if(!u&&a)return a(o,!0);if(i)return i(o,!0);var f=new Error("Cannot find module '"+o+"'");throw f.code="MODULE_NOT_FOUND",f}var l=n[o]={exports:{}};t[o][0].call(l.exports,function(e){var n=t[o][1][e];return s(n?n:e)},l,l.exports,e,t,n,r)}return n[o].exports}var i=typeof require=="function"&&require;for(var o=0;o<r.length;o++)s(r[o]);return s})({1:[function(require,module,exports){
var vector = require('./vector');
var optimize = module.exports = {};

function q(x) {
    // The existential operator;
    return typeof(x) !== "undefined" && x !== null;
}

function max_abs_diff(x0, x) {
    var max = 0.0;

    for (var i = 0; i < x.length; i++){
        max = Math.max(max, Math.abs(x0 - x[i]));
    }
    return max;
}

optimize.fmin = function (func, x0, opts) {
    // Optimize a function using Nelder-Mead.

    x0 = vector.atLeast1d(x0);
    var N = x0.length;

    // Defaults.
    if (!q(opts)){ opts = {}; }

    // Not used
    //var xtol = q(opts.xtol) ? opts.xtol : 1e-6;

    var ftol = q(opts.ftol) ? opts.ftol : 1e-6;
    var maxiter = q(opts.maxiter) ? opts.maxiter : 200 * N;

    // Magic numbers from `scipy`.
    var rho = 1;
    var chi = 2;
    var psi = 0.5;
    var sigma = 0.5;
    var nonzdelt = 0.05;
    var zdelt = 0.00025;

    var sim = [];
    sim[0] = x0;
    var fsim = [];
    fsim[0] = func(x0);

    for (var i = 0; i < N; i++) {
        var y = vector.copy(x0);
        if (y[i] !== 0.0){
            y[i] *= (1 + nonzdelt);
        }else {
            y[i] = zdelt;
        }

        sim[i + 1] = y;
        fsim[i + 1] = func(y);
    }

    var inds = vector.argsort(fsim);
    fsim = vector.take(fsim, inds);
    sim = vector.take(sim, inds);

    var iterations = 0;

    // Constraint on function calls is needed.
    while (iterations < maxiter) {
        var doshrink = false;

        iterations += 1;
        // A break based on xtol needs to be included too.
        if (max_abs_diff(fsim[0], fsim.slice(1, fsim.length)) <= ftol){
            break;
        }

        var xbar = vector.fmult(1.0 / N,
            vector.reduce(sim.slice(0, sim.length - 1)));
        var xr = vector.add(vector.fmult(1 + rho, xbar),
            vector.fmult(-rho, sim[sim.length - 1]));
        var fxr = func(xr);

        if (fxr < fsim[0]) {
            var xe = vector.add(vector.fmult(1 + rho * chi, xbar),
                vector.fmult(-rho * chi, sim[sim.length - 1]));
            var fxe = func(xe);
            if (fxe < fxr) {
                sim[sim.length - 1] = xe;
                fsim[fsim.length - 1] = fxe;
            } else {
                sim[sim.length - 1] = xr;
                fsim[fsim.length - 1] = fxr;
            }
        } else {
            if (fxr < fsim[fsim.length - 2]) {
                sim[sim.length - 1] = xr;
                fsim[fsim.length - 1] = fxr;
            } else {
                var xc, fxc;
                if (fxr < fsim[fsim.length - 1]) {
                    xc = vector.add(vector.fmult(1 + rho * psi, xbar),
                        vector.fmult(-rho * psi, sim[sim.length - 1]));
                    fxc = func(xc);
                    if (fxc < fsim[fsim.length - 1]) {
                        sim[sim.length - 1] = xc;
                        fsim[fsim.length - 1] = fxc;
                    } else {
                        doshrink = true;
                    }
                } else {
                    xc = vector.add(vector.fmult(1 - psi, xbar),
                        vector.fmult(psi, sim[sim.length - 1]));
                    fxc = func(xc);
                    if (fxc < fsim[fsim.length - 1]) {
                        sim[sim.length - 1] = xc;
                        fsim[fsim.length - 1] = fxc;
                    } else {
                        doshrink = true;
                    }
                }

                if (doshrink) {
                    for (var j = 1; j < N + 1; j++) {
                        sim[j] = vector.add(sim[0], vector.fmult(sigma,
                            vector.subtract(sim[j], sim[0])));
                        fsim[j] = func(sim[j]);
                    }
                }
            }
        }

        inds = vector.argsort(fsim);
        fsim = vector.take(fsim, inds);
        sim = vector.take(sim, inds);
    }

    var x = sim[0];
    var fval = fsim[0];

    var result = {};

    if (iterations >= maxiter){
        result.error = "Too many iterations. " + iterations;
    }else{
        result.result = "Converged in " + iterations + " iterations.";
    }

    result.x = x;
    result.fval = fval;

    return result;
};

},{"./vector":2}],2:[function(require,module,exports){
var vector = module.exports = {};

function merge(l, r, data) {
    // Merging for use with a merge argsort.
    var result = [];
    while (l.length && r.length) {
        if (data[l[0]] <= data[r[0]]){
            result.push(l.shift());
        }else{
            result.push(r.shift());
        }
    }
    while (l.length){
        result.push(l.shift());
    }
    while (r.length){
        result.push(r.shift());
    }
    return result;
}

vector.shape = function(arr, ret){
    if(!ret){ ret = []; }

    if(Array.isArray(arr)){
        ret.push(arr.length);
        return vector.shape(arr[0], ret);
    }else{
        return ret;
    }
};

vector.copy = function (x) {
    var y, i;
    if (typeof(x.length) === "undefined"){
        return x;
    }

    y = [];
    for (i = 0; i < x.length; i++){
        y[i] = x[i];
    }

    return y;
};

vector.atLeast1d = function (x) {
    // Make sure that an object acts as an array (even if it's a scalar).
    if (typeof(x.length) === "undefined") {
        var tmp = [];
        tmp[0] = x;
        return tmp;
    }
    return x;
};

vector.range = function (a, b, c) {
    var xmin, xmax, dx, x, rng = [];
    if (typeof(b) === "undefined") {
        xmin = 0;
        xmax = a;
        dx = 1;
    } else if (typeof(c) === "undefined") {
        xmin = a;
        xmax = b;
        dx = 1;
    } else {
        xmin = a;
        xmax = b;
        dx = c;
    }

    var i;
    for (x = xmin, i = 0; x < xmax; x += dx, i++){
        rng[i] = x;
    }

    return rng;
};

vector.dot = function (a, b) {
    var i, result = 0.0;
    if (a.length !== b.length){
        throw "Size mismatch in vector.dot.";
    }

    for (i = 0; i < a.length; i++) {
        result += a[i] * b[i];
    }
    return result;
};

vector.fmult = function (f, v) {
    var i, result = [];
    for (i = 0; i < v.length; i++) {
        result[i] = f * v[i];
    }
    return result;
};

vector.add = function (a, b) {
    var i, result = [];
    if (a.length !== b.length) {
        throw "Size mismatch in vector.add.";
    }
    for (i = 0; i < a.length; i++) {
        result[i] = a[i] + b[i];
    }
    return result;
};

vector.subtract = function (a, b) {
    var i, result = [];
    if (a.length !== b.length) {
        throw "Size mismatch in vector.subtract.";
    }
    for (i = 0; i < a.length; i++) {
        result[i] = a[i] - b[i];
    }
    return result;
};

vector.reduce = function (x) {
    var i, result = x[0];
    for (i = 1; i < x.length; i++){
        result = vector.add(result, x[i]);
    }
    return result;
};

vector.take = function (x, ind) {
    // Re-order a vector.
    var i, result = [];
    for (i = 0; i < ind.length; i++) {
        result[i] = x[ind[i]];
    }
    return result;
};

function rec_argsort(inds, data) {
    // The recursive helper function for the argsort.
    var m, l, r;
    if (typeof(inds) === "undefined" || inds.length === 1) {
        return inds;
    }

    m = parseInt(inds.length / 2);
    l = inds.slice(0, m);
    r = inds.slice(m, inds.length);
    return merge(rec_argsort(l, data),
        rec_argsort(r, data), data);
}

vector.argsort = function (x) {
    // Argsort of an array using merge sort.
    if (typeof(x.length) === "undefined" || x.length === 1) {
        return x;
    }

    return rec_argsort(vector.range(x.length), x);
};

},{}]},{},[1]);
